engine/new_runtime/src/memories/noncontiguous_memory.rs

```diff
@@ -0,0 +1,342 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    locked_memory::LockedMemory,
+    memories::{buffer::Buffer, file_memory::FileMemory, ram_memory::RamMemory},
+    utils::*,
+    MemoryError::*,
+    *,
+};
+use core::{
+    fmt::{self, Debug, Formatter},
+    marker::PhantomData,
+};
+use crypto::hashes::sha;
+use zeroize::Zeroize;
+
+use serde::{
+    de::{Deserialize, Deserializer, SeqAccess, Visitor},
+    ser::{Serialize, Serializer},
+};
+
+static IMPOSSIBLE_CASE: &str = "NonContiguousMemory: this case should not happen if allocated properly";
+
+// Currently we only support data of 32 bytes in noncontiguous memory
+pub const NC_DATA_SIZE: usize = 32;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum NCConfig {
+    FullFile,
+    FullRam,
+    RamAndFile,
+}
+use NCConfig::*;
+
+// NONCONTIGUOUS MEMORY
+/// Shards of memory which composes a non contiguous memory
+#[derive(Clone)]
+enum MemoryShard {
+    FileShard(FileMemory),
+    RamShard(RamMemory),
+}
+use MemoryShard::*;
+
+/// NonContiguousMemory only works on data which size corresponds to the hash primitive we use. In our case we use it to
+/// store keys hence the size of the data depends on the chosen box provider
+#[derive(Clone)]
+pub struct NonContiguousMemory {
+    shard1: MemoryShard,
+    shard2: MemoryShard,
+    config: NCConfig,
```

Comment:
I don't think we need the `config` property here, it is only used on `alloc`. And if we need it again, we can just derive it by matching on the variants of `shard1` and `shard2`.